### ISSUE TITLE
bug in TimeStepManager?

### ISSUE THREAD

**Issue by ecoon** [amanzi/amanzi#685]
This line of code seems suspicious to me:  https://github.com/amanzi/amanzi/blob/master/src/utils/TimeStepManager.cc#L122

My understanding is that this branch of the code is intended to avoid small "next" timesteps.  So for instance, if the time til the next "event" is 1s, and the timestep is 0.9 seconds, we would rather take two 0.5-second steps than a 0.9 + a 0.1.

The return statement looks right to me, e.g. https://github.com/amanzi/amanzi/blob/master/src/utils/TimeStepManager.cc#L127

I'm confused by what is being reserved here, and why.  

In this step, `dt_stable_storage = 2*dT - 0.5*time_remaining` (line 122).   We know that `dT > 0.75*time_remaining` (the condition), meaning that `2*dT > 1.5*time_remaining` meaning that `2*dT - 0.5*time_remaining = dt_stable_storage > time_remaining`. 

Then, at the next timestep, dt_stable_storage will always be values that ought to be limited by the next event (as time_remaining will now be half the previous value).  dT is assigned to be dt_stable_storage, and we will always hit the dT > time_remaining condition branch (line 113).  Now dt_stable_storage is assigned to be dT (this doesn't change anything) and we continue with dt as the time remaining, as expected.

Finally, in the next timestep (the first step after the event), dT is assigned, once again, to dt_stable_storage.  Presumably this is not limited by an event, so we will take this step, with a value of `2*dT - 0.5*time_remaining.`

My concern here is whether this is valid for the PK?  In particular, one would expect, since `dT > 0.75*time_remaining`, clearly `2*dT - 0.5*time_remaining > dT`.  So we're asking the PK to take a step larger than the one it last said was valid. 

So there are two questions here really:

1. Is this code correct?  Am I missing something?  I definitely think we should not ask the PK to take a step more than epsilon bigger than what it has said is valid, and this is definitely bigger than that.
2. Is storing dt_stable_timestep code necessary?  I believe the PK should be responsible for storing its timestep, and in the case where an event limits the actual timestep to less than is valid for the PK, and the PK happily solves that timestep with no trouble (not surprisingly), the PK can keep asking for bigger timestep sizes.  This is currently implemented in ATS.
3. If it is necessary, shouldn't this just get set to dT at line 122?  This will follow the same path through the code, and in the first step "after" the event, we would use the saved value, dT.

I recently added an assertion in ATS code that checks that the timestep prescribed by the coordinator to a PK is never bigger than the PK's "valid" timestep (returned via `PK::get_dt()`).  This assertion is now throwing in a very few set of test problems.

**Comment by dasvyat** [comment]
My understanding that the algorithm tries to avoid approaching to close to event times. If the event is in 0.9 s and your dt is 1s then the time stepping is 0.45s , 0.45s and 1s. The algorithm should work like this

1. Time step is cut 0.45s,  dt_stable_storage = 1
2. dT = 1s, but then cut again dT = 0.45s and dt_stable_storage = 1s. Event is reached
3. dT = 1s

**Comment by ecoon** [comment]
@dasvyat  That's not right though -- if dT is 1s and the event is in 0.9 s, then 

1. Time step is cut to 0.9, dt_stable_storage = 1.  Event is reached.
2. dT = 1s

That part is implemented correctly.

Maybe you meant to reverse those -- if dT is 0.9s and the event is in 1s, then what _currently_ happens is:

1. Time step is cut to 0.5, dt_stable_storage = 2*dT-0.5*time_remaining = 1.8-0.5 = 1.3
2. dT from PK is 0.9, this is overwritten with dt_stable_storage to 1.3, but then cut by event to 0.5.  Event is reached.
3. dT from PK is 0.9, this is overwritten with dt_stable_storage to 1.3.   ATS now asserts here because PK never said we could go bigger than 0.9...

I think that it _should_ be:
1. Time step is cut to 0.5, dt_stable_storage = dT = 0.9
2. dT is 0.9, but then cut by event to 0.5 and dt_stable_storage=0.9  Event is reached.
3. dT is 0.9

Are we in agreement with that?

**Comment by dasvyat** [comment]
Yes, you are right that I meant the case you described above when dT is 0.9s and the event is in 1s.
I agree with you that this is wrong and TimeStepManager should not increase it above stable. dt_stable_storage = dT

**Comment by ecoon** [comment]
In a conversation with @lipnikov  and @jd-moulton  we agreed on the following:

1. the old, stable timestep size should be saved by the TimeStepController.  Other choices include: 1, the TimeStepManager, which is bad because, in a case with multiple PKs and/or multiple MPCs with their own TimeStepController, we will only ever recover the timestep to the min of the old values, and 2, the PK, which is bad because that puts more logic in the PK instead of keeping the timestep size logic in the TimeStepController.
2. TimeStepManager is a confusing name that is too close to TimeStepController, and we should rename this to something like EventManager or IOEventManager since it accumulates the needs of all IOEvents.
3. A quick fix of dt_stable_storage=dT is fine for now, but we can leave this ticket to deal with #1.
4. When we do #1, we can remove the logic in ATS in pk_bdf_default which deals with this case.

**Comment by lipnikov** [comment]
Another bug: Time step manager increases physics time step. I think this logic was added by @ecoon 

```
CycleDriver        |  Cycle 939: time = 0.098931 s, dt = 4.24988e-05 s
ShallowWater       |   min/max(h): 0/5.9726, min/max(flux): -31.3357/30.1238
Transport-domain   |   1 sub-cycles, dt_stable=0.000116113 s, dt_MPC=4.24988e-05 s
TimeStepManager    |   Proposed dT=4.11358e-05 s, is near equal to next event time remaining 0.00102653 s.
CycleDriver        |  New time = 0.0989735 s
CycleDriver        |
CycleDriver        |  Cycle 940: time = 0.0989735 s, dt = 0.00102653 s
```

**Comment by lipnikov** [comment]
@ecoon  Assumption that small increment does not play a role sounds fair but may lead to small overshoots in explicit schemes. However, since PK suggests dt, I do not want to say No and suggest to use dt + eps, even is eps is relative. If you want to keep this if(), let us make is optional via control flag.

**Comment by ecoon** [comment]
Fair enough.  As long as we have the option to take two half-steps, I'm probably ok with eliminating that branch of the code completely.  At least we can try it and see if there are any corner cases where we get roundoff error problems.

**Comment by ecoon** [comment]
So despite fixes in #694, there is still a bug here that I think probably can't be fixed until we deal with #1.

Effectively I have a problem where the following happens:

- Time events happen every day.
- The code takes steps of size 0.4 days.
- Step 1 is good, Step 2 is good, so we are now at 0.8 days.
- Step 3, the PK says 0.4 days, but the step is limited by the TimeStepManager to 0.2 days to match the 1-day event.
- Despite the fact that the PK thought it could take 0.4 days, some nonlinearity kicks in and the 0.2 day step takes 19 nonlinear iterations (just less than the "limit iterations" of 20).  Therefore the PK says, oh, cut the timestep for the next step.
- Therefore the PK requests the step to be 0.1 days.  But the TimeStepManager recalls that it limited the step size, and "recovers" the old step size of 0.4, overriding the  PK.
